src/styles.js

```diff
@@ -82,6 +72,58 @@ export function installStyles(doc, cssText, cb, opt_isRuntimeCss, opt_ext) {
   }, 4);
 }
 
+
+/**
+ * Adds the given css text to the given shadow root.
+ *
+ * The style tags will be at the beginning of the shadow root before all author
+ * styles. One element can be the main runtime CSS. This is guaranteed
+ * to always be the first stylesheet in the doc.
+ *
+ * @param {!ShadowRoot} shadowRoot
+ * @param {string} cssText
+ * @param {boolean=} opt_isRuntimeCss If true, this style tag will be inserted
+ *     as the first element in head and all style elements will be positioned
+ *     after.
+ * @param {string=} opt_ext
+ */
+export function installStylesForShadowRoot(shadowRoot, cssText,
+    opt_isRuntimeCss, opt_ext) {
+  insertStyleElement(
+      shadowRoot.ownerDocument,
+      shadowRoot,
+      cssText,
+      opt_isRuntimeCss || false,
+      opt_ext || null);
+}
+
+
+/**
+ * Creates the properly configured style element.
+ * @param {!Document} doc
+ * @param {!Element|!ShadowRoot} cssRoot
+ * @param {string} cssText
+ * @param {boolean} isRuntimeCss
+ * @param {?string} ext
+ * @return {!Element}
+ */
+function insertStyleElement(doc, cssRoot, cssText, isRuntimeCss, ext) {
+  const style = doc.createElement('style');
+  style.textContent = cssText;
+  let afterElement = null;
+  // Make sure that we place style tags after the main runtime CSS. Otherwise
+  // the order is random.
+  if (isRuntimeCss) {
+    style.setAttribute('amp-runtime', '');
+  } else {
+    style.setAttribute('amp-extension', ext || '');
+    afterElement = cssRoot.querySelector('style[amp-runtime]');
```

Comment:
Cache this value in cssRoot instead of running selector engine.

Don't have to do right now, of course.